File: apps/ios/Loonext/Core/ScheduledSend.swift
import Foundation

/// #233 — send later, as this phone understands it.
///
/// A hand-port of packages/shared/src/scheduled-send.ts, mirrored again in
/// android/core/scheduled/ScheduledSend.kt. What is shared is the SPEC: which
/// presets exist, the hour they land on, the order, the wording of every reason
/// a message did not go, and which of those reasons will clear on their own.
///
/// The reasons matter more here than the timing does. `docs/DECISIONS.md` makes
/// disclosure binding — held not dropped, everything held or cancelled told to
/// the owner, time-sensitive work expiring rather than arriving late — and a
/// rule about disclosure is only as good as the sentence doing the disclosing.
/// Three clients writing their own version of "we did not send this" is how one
/// of them ends up saying nothing at all.
///
/// The calendar arithmetic is deliberately NOT ported. It uses Foundation here
/// and java.time on Android because only a real calendar gets DST right, the
/// same reason `SnoozeLogic.swift` gives.
///
/// Resolved in the DESTINATION's zone rather than the device's, which is the
/// one place this differs from the snooze ladder: a snooze is when YOU want to
/// see something again, and a scheduled text is when THEY read it.
enum ScheduledSend {
    /// The hour presets land on. Early enough to be first in the inbox.
    static let presetHour = 8

    /// The longest a scheduled body may be. Mirrors the column check.
    static let bodyMax = 1600

    /// How far out a send may be scheduled. Mirrors the SQL horizon.
    static let horizonDays = 90

    /// How many live scheduled messages one workspace may hold. Mirrors SQL.
    static let perCompanyCap = 200

    /// ...and one thread, so a conversation cannot become a drip campaign.
    static let perThreadCap = 20

    /// Statuses a scheduled message can be in, mirroring the column's CHECK.
    static let statuses = ["pending", "held", "sent", "canceled", "expired", "failed"]

    /// Still going, as far as anybody knows.
    static func isLive(_ status: String) -> Bool {
        status == "pending" || status == "held"
    }

    /// Why a scheduled message did not go, in the words the owner reads.
    ///
    /// Each is a REASON, not an error. `recipient_opted_out` deliberately
    /// offers no remedy, because there is not one — an opt-out can only be
    /// lifted by the customer, which is carrier truth rather than our policy.
    /// #228: KEYS, not sentences. This map is built at type-init, before any
    /// reader exists.
    private static let holdReasonKeys: [String: String] = [
        "subscription_inactive": "domain.scheduledHoldSubscriptionInactive",
        // #277: the seasonal hold. A SEPARATE reason from a lapse because the
        // events and the remedies are separate: nothing lapsed, no card needs
        // sorting, and the number is not on any clock. The sentence above used
        // to say "paused" for a lapse; that word belongs to this now, and two
        // reasons both claiming it is the confusion this roster exists to stop.
        "workspace_paused": "domain.scheduledHoldWorkspacePaused",
        "registration_pending": "domain.scheduledHoldRegistrationPending",
        "service_unavailable": "domain.scheduledHoldServiceUnavailable",
        "calendar_unverified": "domain.scheduledHoldCalendarUnverified",
        "customer_replied": "domain.scheduledHoldCustomerReplied",
        "recipient_opted_out": "domain.scheduledHoldOptedOut",
        "invalid_destination": "domain.scheduledHoldInvalidDestination",
        "expired": "domain.scheduledHoldExpired",
        "workspace_closed": "domain.scheduledHoldWorkspaceClosed",
        // #237: done, deleted, or reminders switched off for that job. One
        // reason for three causes — from the reader's side the actionable fact
        // is identical, and three near-identical sentences is the drift this
        // roster exists to prevent.
        "job_no_longer_scheduled": "domain.scheduledHoldJobUnscheduled",
    ]

    /// Every reason, resolved, in the reader's language.
    static func localisedHoldReasons(_ locale: String? = nil) -> [String: String] {
        holdReasonKeys.mapValues { AppStrings.translate(locale, $0) }
    }

    /// The English, for the callers and the guards that have no reader.
    ///
    /// `ScheduledSendTests` walks this asserting every reason is a SENTENCE and
    /// not a code, which is a guard worth keeping pointed at the copy rather
    /// than at the keys.
    static var holdReasons: [String: String] { localisedHoldReasons() }

    /// One reason, or nil for a state we do not have words for.
    static func holdReason(_ reason: String, locale: String? = nil) -> String? {
        holdReasonKeys[reason].map { AppStrings.translate(locale, $0) }
    }

    /// Does this reason clear on its own?
    ///
    /// Drives whether the UI offers "we will keep trying" or asks for a
    /// decision. A reason wrongly marked recoverable is a message that retries
    /// forever against a condition that will never change.
    static func reasonRecovers(_ reason: String) -> Bool {
        switch reason {
        // #277: a pause is the most recoverable state in the product. It is a
        // season, and the whole promise is that everything is where it was left
        // when the crew comes back. Marked terminal, pausing would quietly
        // destroy a workspace's scheduled work.
        case "subscription_inactive", "workspace_paused", "registration_pending",
             "service_unavailable", "calendar_unverified", "customer_replied":
            return true
        default:
            return false
        }
    }

    /// The sentences the send-later UI says on every client.
    ///
    /// `holdReasons` covers the states where a message did NOT go; this covers
    /// the rest of the surface — the picker, the quiet-hours warning, the
    /// confirmations. Here for the same reason: three clients writing their own
    /// version of "that lands late where they are" is three different products,
    /// and the phone is where somebody schedules a text at 9:40pm with the van
    /// still running.
    ///
    /// Whole sentences only. Button labels stay per-platform, because a
    /// SwiftUI `Button` role and a web dialog footer have different conventions
    /// and a shared "Cancel" would be pretending otherwise.
    private static let copyKeys: [String: String] = [
        "picker_reassurance": "domain.scheduledPickerReassurance",
        "quiet_hours_choice": "domain.scheduledQuietHoursChoice",
        "quiet_hours_unknown": "domain.scheduledQuietHoursUnknown",
        "canceled_confirmation": "domain.scheduledCancelled",
        "nothing_scheduled": "domain.scheduledNothingWaiting",
    ]

    /// The English, for the callers that have no reader.
    static var copy: [String: String] {
        copyKeys.mapValues { AppStrings.translate(nil, $0) }
    }

    /// One line of ``copy``, or empty rather than a crash on a key typo.
    ///
    /// The empty string is deliberate and unchanged: a typo'd key must leave a
    /// gap in a sentence somebody notices, not take a screen down.
    static func copyLine(_ key: String, locale: String? = nil) -> String {
        copyKeys[key].map { AppStrings.translate(locale, $0) } ?? ""
    }

    /// Whose clock the sender picked against, said out loud.
    ///
    /// The same three rungs and the same wording as the thread's "their time"
    /// line (`clockProvenance` in MessagingRepository.swift) — a product that
    /// says "from their area code" in one place and something else in another
    /// has two vocabularies for one fact.
    /// #228 — why a held message did not go, in the reader's language.
    ///
    /// Hand-port of `scheduledHoldText` in packages/shared. The API writes BOTH
    /// a catalogue key and the English sentence it has always written, so:
    /// translate the key where we have words for it, and otherwise render the
    /// stored sentence — which is what a row written before the key existed has.
    ///
    /// A key with no words resolves to ITSELF (the catalogue fails open), so a
    /// self-resolving key counts as absent. Without that check a French reader
    /// would see `domain.scheduledHoldExpired` rather than an English sentence,
    /// which is the worse of the two.
    static func holdText(
        reasonKey: String?,
        storedEnglish: String?,
        locale: String? = nil
    ) -> String? {
        if let reasonKey, !reasonKey.isEmpty {
            let words = AppStrings.translate(locale, reasonKey)
            if !words.isEmpty, words != reasonKey { return words }
        }
        return storedEnglish
    }

    static func clockProvenance(_ source: String, locale: String? = nil) -> String {
        switch source {
        case "contact": return AppStrings.translate(locale, "domain.clockTheirTimeContact")
        case "area_code": return AppStrings.translate(locale, "domain.clockTheirTimeAreaCode")
        default: return AppStrings.translate(locale, "domain.clockWorkspaceTime")
        }
    }
}

/// One offer in the send-later menu. `at` is nil for the picker.
struct SchedulePreset: Equatable {
    let id: String
    let label: String
    let at: Date?
}

/// Days from `date` forward to the next Monday — never 0, always next week.
///
/// Deliberately a copy of `daysUntilNextMonday` in SnoozeLogic.swift rather
/// than a call to it: that one is the snooze ladder's, resolved in the DEVICE's
/// calendar, and this one takes the destination's. Sharing the function would
/// mean one of the two features silently changing zone the day somebody
/// "simplified" the other.
///
/// PREFIXED, not just `private`. A top-level `private` func still occupies the
/// module's namespace, so this and SnoozeLogic's identical signature were an
/// "invalid redeclaration" — a break only CI's iOS job could see, because
/// nothing on this side of the repo compiles Swift. The name now says which
/// feature's calendar it is, which is what the doc above was already for.
private func sendLaterDaysUntilNextMonday(_ date: Date, calendar: Calendar) -> Int {
    // Calendar's weekday is 1 = Sunday … 7 = Saturday, which is NOT the same
    // numbering as java.time's DayOfWeek or JavaScript's getDay(). Converting
    // to a Monday-is-1 index first keeps this identical to the other two
    // clients rather than merely similar.
    let weekday = calendar.component(.weekday, from: date)
    let mondayBased = weekday == 1 ? 7 : weekday - 1  // Mon = 1 … Sun = 7
    return mondayBased == 7 ? 1 : 8 - mondayBased
}

/// The two presets plus the escape hatch, in the DESTINATION's zone.
///
/// Two, not five: #233 names exactly these, and a preset list long enough to
/// need reading is slower than the picker it was meant to avoid.
func schedulePresets(
    now: Date = Date(),
    timeZone: TimeZone,
    locale: String? = nil
) -> [SchedulePreset] {
    var calendar = Calendar(identifier: .gregorian)
    calendar.timeZone = timeZone

    func at(addDays: Int) -> Date? {
        guard let day = calendar.date(byAdding: .day, value: addDays, to: now) else {
            return nil
        }
        return calendar.date(
            bySettingHour: ScheduledSend.presetHour,
            minute: 0,
            second: 0,
            of: day,
            matchingPolicy: .nextTime
        )
    }

    return [
        SchedulePreset(
            id: "tomorrow",
            label: AppStrings.translate(locale, "domain.scheduledPresetTomorrow"),
            at: at(addDays: 1)
        ),
        SchedulePreset(
            id: "monday",
            label: AppStrings.translate(locale, "domain.scheduledPresetMonday"),
            at: at(addDays: sendLaterDaysUntilNextMonday(now, calendar: calendar))
        ),
        SchedulePreset(
            id: "custom",
            label: AppStrings.translate(locale, "domain.scheduledPresetCustom"),
            at: nil
        ),
    ]
}

File: apps/api/src/messaging/scheduled-send.ts
/**
 * #233 — the job that turns a scheduled intent into a text.
 *
 * Runs every minute alongside the lead-chase ladder, on the same argument: the
 * scan is a partial index over due rows only, so a quiet minute costs one
 * indexed lookup returning nothing, and a coarser cadence would make "8:00am"
 * mean "some time between 8:00 and 8:05".
 *
 * ---------------------------------------------------------------------------
 * THE GATES RUN NOW, NOT WHEN IT WAS SCHEDULED
 *
 * This is the whole reason the feature is not a client-side timer. Between
 * somebody writing a text on Friday night and it going out on Monday morning,
 * the customer can send STOP, the workspace's card can fail, carrier
 * registration can lapse, and the number can be released. So the firing path
 * calls `runPreSendGates` exactly like an immediate send, and hands the
 * resulting clearance to `dispatchOutbound`.
 *
 * That is not a convention this file could forget: #331 made the clearance a
 * value only `runPreSendGates` can mint, and `dispatchOutbound` demands one, so
 * a send path that skipped the gates would not compile. The opt-out guarantee
 * in #233's acceptance criteria is satisfied by construction.
 *
 * ---------------------------------------------------------------------------
 * HELD IS NOT FAILED, AND BOTH ARE DISCLOSED
 *
 * `docs/DECISIONS.md` fixed this before the feature existed, against #325:
 * held not dropped, everything held or cancelled disclosed to the owner when it
 * happens, and time-sensitive work expiring rather than arriving late.
 *
 * The distinction is load-bearing rather than cosmetic. A HOLD means the
 * condition can clear and the message will go — a paused subscription, a
 * pending registration, an incident. A FAILURE means it never will, and saying
 * "we'll keep trying" would be a lie: an opt-out can only be lifted by the
 * customer, and a number we cannot text will not become textable by waiting.
 * `scheduledReasonRecovers` in @loonext/shared owns that classification so the
 * copy and the retry behaviour cannot disagree.
 *
 * ---------------------------------------------------------------------------
 * TALKING OVER SOMEBODY
 *
 * The one hold that is about manners rather than capability. If the customer
 * replied after this was scheduled, "still thinking about that quote?" arrives
 * after they already said yes, which reads as a robot. The row carries the
 * conversation's newest inbound at scheduling time; a newer one here means the
 * thread moved on, so it holds and asks rather than sending.
 *
 * Deliberately a HOLD and not a cancel: the owner may well still want to send
 * it, and deciding that for them is how a message silently disappears.
 */
import {
  SCHEDULED_HOLD_REASONS,
  SCHEDULED_HOLD_REASON_KEYS,
  type ScheduledHoldReason,
  estimateSegments,
  scheduledReasonRecovers,
} from "@loonext/shared";
import type { SupabaseClient } from "@supabase/supabase-js";

import { listConversationViewers } from "../auth/conversation-audience";
import { CALENDAR_VERIFICATION_MAX_AGE_MS } from "../calendar/liveness";
import { getDb } from "../db";
import type { Env } from "../env";
import { ApiError } from "../http/errors";
import { deliverPush } from "../notifications/deliver";
import { SCHEDULED_DISCLOSURE_COPY } from "./scheduled-send-copy";
import { dispatchOutbound, runPreSendGates } from "./send";
import type { MessageRow } from "./types";

/**
 * One run's ceiling. Comfortably above any plausible minute — the per-workspace
 * cap is 200 live rows in total, so this only binds if many workspaces happen
 * to schedule the same minute, and the remainder simply goes on the next tick.
 */
const SCAN_LIMIT = 100;

/**
 * How long a claim is good for.
 *
 * Long enough that the gates and a carrier round trip finish inside it, short
 * enough that a worker killed mid-flight does not leave a customer's message
 * unsent for an hour. Five minutes matches the interrupted-send sweep that
 * owns the window after this one.
 */
const LEASE_SECONDS = 300;

interface ScheduledRow {
  id: string;
  company_id: string;
  conversation_id: string;
  body: string;
  send_at: string;
  inbound_watermark: string | null;
  /** #237: the job this reminds about, or null for a text a person wrote. */
  task_id: string | null;
  origin: string;
}

interface Destination {
  from: string;
  to: string;
  newestInbound: string | null;
}

type FireScheduledMessageResult =
  | { outcome: "fired"; message: MessageRow }
  | { outcome: "gone" }
  | {
      outcome: "held";
      reason_key: string;
      scheduled_message?: ScheduledRow;
    }
  | {
      outcome: "failed";
      reason_key: string;
      scheduled_message?: ScheduledRow;
    };

export interface ScheduledSendSummary {
  sent: number;
  held: number;
  failed: number;
  expired: number;
}

function unwrap<T>(
  result: { data: unknown; error: { message: string } | null },
  what: string,
): T {
  if (result.error) throw new Error(`${what} failed: ${result.error.message}`);
  return result.data as T;
}

/**
 * Which gate failure this was, in terms the owner can be told about.
 *
 * `runPreSendGates` throws `ApiError` with a code, and the mapping is
 * exhaustive over the five it can raise. An unrecognised code becomes a
 * SERVICE_UNAVAILABLE hold rather than a failure, deliberately: a gate this
 * file has not been taught about is our ignorance, and holding a message we do
 * not understand is recoverable while discarding it is not.
 */
function reasonFor(cause: unknown): ScheduledHoldReason {
  if (!(cause instanceof ApiError)) return "service_unavailable";
  switch (cause.code) {
    case "recipient_opted_out":
      return "recipient_opted_out";
    case "subscription_inactive":
      return "subscription_inactive";
    // #277: held, not failed, and with its own sentence — a paused workspace's
    // scheduled sends go out when they resume, which is the whole point of
    // pausing rather than cancelling.
    case "workspace_paused":
      return "workspace_paused";
    case "registration_pending":
      return "registration_pending";
    case "validation_failed":
      return "invalid_destination";
    default:
      return "service_unavailable";
  }
}

/** The numbers this send is between, and where the thread actually stands. */
async function resolveDestination(
  db: SupabaseClient,
  row: ScheduledRow,
): Promise<Destination | null> {
  const { data, error } = await db
    .from("conversations")
    // #291: the thread's number, resolved at SEND time. A number
    // stamped when the text was scheduled would go stale if the crew
    // corrected it in between.
    .select("contact_phone_e164,phone_numbers(number_e164,status)")
    .eq("id", row.conversation_id)
    .eq("company_id", row.company_id)
    .maybeSingle();
  if (error) throw new Error(`scheduled destination lookup: ${error.message}`);

  const conversation = data as unknown as {
    contact_phone_e164: string | null;
    phone_numbers: { number_e164: string; status: string } | null;
  } | null;

  const to = conversation?.contact_phone_e164;
  const number = conversation?.phone_numbers;
  // A released number cannot be sent from, and the gates would not catch it —
  // they check the workspace, not which of its numbers this thread is on.
  if (!to || !number || number.status === "released") return null;

  const { data: newest, error: newestError } = await db
    .from("messages")
    .select("created_at")
    .eq("company_id", row.company_id)
    .eq("conversation_id", row.conversation_id)
    .eq("direction", "inbound")
    .order("created_at", { ascending: false })
    .limit(1)
    .maybeSingle();
  if (newestError) {
    throw new Error(`scheduled inbound watermark: ${newestError.message}`);
  }

  return {
    from: number.number_e164,
    to,
    newestInbound: (newest as { created_at: string } | null)?.created_at ?? null,
  };
}

/** Did the customer answer after this was written? */
function customerReplied(row: ScheduledRow, destination: Destination): boolean {
  if (!destination.newestInbound) return false;
  // No watermark means the thread had no inbound when this was scheduled, so
  // any inbound now is newer by definition.
  if (!row.inbound_watermark) return true;
  return (
    new Date(destination.newestInbound).getTime() >
    new Date(row.inbound_watermark).getTime()
  );
}

/**
 * #237 — is the job this reminder is about still on the books?
 *
 * THE GUARANTEE, AS OPPOSED TO THE OPTIMISATION. Regenerating a job's reminders
 * already removes them when it changes, and that is what keeps the thread strip
 * and the workspace list accurate. But it depends on every write path calling
 * the sync, and "did this edit change whether the job deserves reminders" is
 * exactly the judgement a caller will one day get wrong — silently, and only
 * visibly when a customer is told to expect somebody who is not coming.
 *
 * So the promise lives HERE, at fire time, where #233 already put every other
 * one. A reminder for a job that is done, deleted, or has reminders switched
 * off does not send, whatever the queue says.
 *
 * Completion DERIVES from the source message's `done_at` (D17) — `tasks` has no
 * done column, and asking for one 400s at PostgREST.
 *
 * Returns true when the row is not a reminder at all: a text a person wrote is
 * theirs to send, and no job's state has any bearing on it.
 */
async function jobStillBooked(
  db: SupabaseClient,
  row: ScheduledRow,
): Promise<boolean> {
  if (row.origin !== "reminder" || !row.task_id) return true;

  const { data, error } = await db
    .from("tasks")
    .select("deleted_at,reminders_off,due_at,messages!message_id(done_at)")
    .eq("id", row.task_id)
    .eq("company_id", row.company_id)
    .maybeSingle();
  // A read that FAILED is not evidence the job is gone. Sending is the
  // recoverable direction here — the job was booked when this was queued, and
  // a transient PostgREST error must not silently cancel somebody's reminder.
  if (error) return true;
  if (!data) return false;

  const task = data as unknown as {
    deleted_at: string | null;
    reminders_off: boolean;
    due_at: string | null;
    // PostgREST types a to-one embed as an ARRAY even where it can only ever
    // hold one row, and returns it as an object. Both shapes are read here
    // rather than picked, because guessing wrong reads as "never done" — and
    // "never done" is the direction that keeps sending.
    messages: { done_at: string | null } | { done_at: string | null }[] | null;
  };
  const source = Array.isArray(task.messages) ? task.messages[0] : task.messages;
  return (
    task.deleted_at === null &&
    !task.reminders_off &&
    task.due_at !== null &&
    (source?.done_at ?? null) === null
  );
}

/**
 * Has every calendar this appointment is mirrored to proved a recent provider
 * round trip?
 *
 * No link means the task is not calendar-backed and keeps the pre-#245 path.
 * Once a live link exists, however, uncertainty fails CLOSED: an unreadable,
 * missing, disconnected, revoked, never-verified, or stale connection holds
 * the reminder. That asymmetry with `jobStillBooked` is deliberate. A failed
 * task read is weighed against the booking that originally queued the reminder;
 * a failed calendar read is exactly the silence D137 says must never be treated
 * as proof that the mirrored time is still right.
 */
async function linkedCalendarsVerifiedRecently(
  db: SupabaseClient,
  row: ScheduledRow,
  now: Date,
): Promise<boolean> {
  if (row.origin !== "reminder" || !row.task_id) return true;

  // A recent provider round trip proves only the last agreed snapshot. A live
  // outbox row means a newer local schedule is still being created, patched,
  // or unlinked, so the literal reminder is not yet safe even when the
  // connection itself is fresh. This check must precede the no-link return:
  // initial provider creates intentionally have no mapping until they commit.
  const { data: outboxData, error: outboxError } = await db
    .from("calendar_outbox")
    .select("id")
    .eq("company_id", row.company_id)
    .eq("task_id", row.task_id)
    .in("state", ["queued", "leased"])
    .limit(1);
  if (outboxError || !Array.isArray(outboxData)) return false;
  if (outboxData.length > 0) return false;

  const { data: linkData, error: linkError } = await db
    .from("task_calendar_links")
    .select("connection_id,link_state")
    .eq("company_id", row.company_id)
    .eq("task_id", row.task_id)
    .neq("link_state", "unlinked");
  if (linkError || !Array.isArray(linkData)) return false;

  const connectionIds: string[] = [];
  for (const link of linkData) {
    const value = link as {
      connection_id?: unknown;
      link_state?: unknown;
    };
    const connectionId = value.connection_id;
    // A conflict, removed event, or refused provider shape means the mirrored
    // appointment is not presently authoritative. Even a fresh connection
    // must not turn that uncertainty into a stale customer reminder.
    if (value.link_state !== "active") return false;
    if (typeof connectionId !== "string" || connectionId.length === 0) {
      return false;
    }
    if (!connectionIds.includes(connectionId)) connectionIds.push(connectionId);
  }
  if (connectionIds.length === 0) return true;

  const { data: connectionData, error: connectionError } = await db
    .from("calendar_connections")
    .select(
      "id,status,last_verified_at,revoked_at,sync_due_at,pull_lease_owner",
    )
    .eq("company_id", row.company_id)
    .in("id", connectionIds);
  if (connectionError || !Array.isArray(connectionData)) return false;

  const connections = new Map(
    connectionData.map((connection) => {
      const value = connection as {
        id: string;
        status: string;
        last_verified_at: string | null;
        revoked_at: string | null;
        sync_due_at: string | null;
        pull_lease_owner: string | null;
      };
      return [value.id, value] as const;
    }),
  );
  const oldestSafeVerification =
    now.getTime() - CALENDAR_VERIFICATION_MAX_AGE_MS;

  return connectionIds.every((connectionId) => {
    const connection = connections.get(connectionId);
    if (
      !connection ||
      connection.status !== "active" ||
      connection.revoked_at !== null ||
      connection.pull_lease_owner !== null ||
      !connection.last_verified_at
    ) {
      return false;
    }
    const verifiedAt = Date.parse(connection.last_verified_at);
    const syncDueAt = connection.sync_due_at
      ? Date.parse(connection.sync_due_at)
      : Number.NaN;
    return (
      Number.isFinite(verifiedAt) &&
      verifiedAt >= oldestSafeVerification &&
      Number.isFinite(syncDueAt) &&
      syncDueAt > now.getTime()
    );
  });
}

/**
 * Tell the workspace a scheduled message is not going, or not going yet.
 *
 * Push only, matching the lead-chase ladder: this is news about work somebody
 * queued and is waiting on, and an email arriving tomorrow about a text that
 * should have gone this morning has already lost. Everyone who can see the
 * thread hears it — a scheduled send belongs to the workspace, and the person
 * who wrote it may be on a roof.
 */
async function disclose(
  env: Env,
  db: SupabaseClient,
  row: ScheduledRow,
  reason: ScheduledHoldReason,
): Promise<void> {
  // Keyed on the thread's NUMBER, not the thread: #106 access is per number,
  // and telling somebody a text is held on a line they cannot see would leak
  // the existence of the conversation. Looked up here rather than carried on
  // the row, because the expiry sweep discloses without resolving a
  // destination first.
  const { data: conversation, error } = await db
    .from("conversations")
    .select("phone_number_id")
    .eq("id", row.conversation_id)
    .eq("company_id", row.company_id)
    .maybeSingle();
  if (error) throw new Error(`scheduled disclosure audience: ${error.message}`);

  const viewers = await listConversationViewers(db, {
    companyId: row.company_id,
    phoneNumberId: (conversation as { phone_number_id: string | null } | null)
      ?.phone_number_id,
  });
  if (viewers.length === 0) return;

  // Held or dead, decided once. The language is decided per reader below; this
  // is not a question about the reader, and asking it inside the composition
  // would be asking it again for every language in the audience.
  const recovers = scheduledReasonRecovers(reason);

  // Push failures are logged, not thrown. The state change has already
  // happened and the thread and the scheduled list both show the row with its
  // reason, so a dead device degrades the disclosure rather than defeating it —
  // whereas failing the run here would leave the NEXT workspace's message
  // unsent because this one's phone was unreachable.
  const failures: unknown[] = [];

  await deliverPush(env, db, {
    category: "messages_mine",
    companyId: row.company_id,
    failures,
    userIds: viewers.map((viewer) => viewer.user_id),
    // #430: every word of this is ours — the reason copy is written in
    // `scheduled-send-copy.ts` over the shared roster, and carries nothing the
    // customer said.
    content: { written: "us" },
    // One identity per scheduled message: a second disclosure about the SAME
    // message replaces the first rather than stacking. A held message is
    // retried every minute, and without this that is a notification a minute.
    collapseKey: `scheduled:${row.id}`,
    // #228: composed per reader. This is the one sentence about a held message
    // that arrives before anybody opens anything, so it is the last place that
    // should still be English inside an otherwise translated app.
    web: (locale) => {
      const copy = SCHEDULED_DISCLOSURE_COPY[locale];
      return {
        title: recovers ? copy.waitingTitle : copy.notSentTitle,
        body: copy.reason[reason],
        url: `${env.APP_ORIGIN}/inbox/${row.conversation_id}`,
      };
    },
  });

  if (failures.length > 0) {
    console.error(
      `scheduled send: ${failures.length} disclosure push(es) failed for ${row.id}`,
    );
  }
}

/** Hold it, and say so. */
async function hold(
  env: Env,
  db: SupabaseClient,
  row: ScheduledRow,
  reason: ScheduledHoldReason,
): Promise<void> {
  unwrap(
    await db.rpc("api_hold_scheduled_message", {
      p_id: row.id,
      // #228: BOTH. The sentence is what a build that predates the key still
      // renders; the key is what a client translates. See the shared module.
      p_reason: SCHEDULED_HOLD_REASONS[reason],
      p_reason_key: SCHEDULED_HOLD_REASON_KEYS[reason],
    }),
    "hold scheduled message",
  );
  await disclose(env, db, row, reason);
}

/** Fail it permanently, and say why. */
async function fail(
  env: Env,
  db: SupabaseClient,
  row: ScheduledRow,
  reason: ScheduledHoldReason,
): Promise<void> {
  unwrap(
    await db.rpc("api_fail_scheduled_message", {
      p_id: row.id,
      p_reason: SCHEDULED_HOLD_REASONS[reason],
      p_reason_key: SCHEDULED_HOLD_REASON_KEYS[reason],
    }),
    "fail scheduled message",
  );
  await disclose(env, db, row, reason);
}

/**
 * Everything past its horizon, expired and disclosed.
 *
 * Runs BEFORE the claim, so a message whose window closed while it was held
 * expires this tick instead of being attempted once more and expiring next
 * time — which would send it late, exactly what rule 3 forbids.
 */
async function expireStale(
  env: Env,
  db: SupabaseClient,
  now: Date,
): Promise<number> {
  const expired = unwrap<ScheduledRow[]>(
    await db.rpc("api_expire_scheduled_messages", {
      p_now: now.toISOString(),
      p_limit: SCAN_LIMIT,
    }),
    "expire scheduled messages",
  );

  for (const row of expired) {
    await disclose(env, db, row, "expired");
  }
  return expired.length;
}

export async function runScheduledSendJob(
  env: Env,
  now: Date = new Date(),
  db: SupabaseClient = getDb(env),
): Promise<ScheduledSendSummary> {
  const summary: ScheduledSendSummary = {
    sent: 0,
    held: 0,
    failed: 0,
    expired: 0,
  };

  summary.expired = await expireStale(env, db, now);

  const due = unwrap<ScheduledRow[]>(
    await db.rpc("api_claim_due_scheduled_messages", {
      p_now: now.toISOString(),
      p_limit: SCAN_LIMIT,
      p_lease_seconds: LEASE_SECONDS,
    }),
    "claim due scheduled messages",
  );

  // Collected rather than thrown one at a time, so one workspace's broken
  // send does not stop every other workspace's from going out this minute
  // (#387). Thrown at the end so a broken run is still visible in Sentry.
  const failures: string[] = [];

  for (const row of due) {
    try {
      const destination = await resolveDestination(db, row);
      if (!destination) {
        await fail(env, db, row, "invalid_destination");
        summary.failed += 1;
        continue;
      }

      if (customerReplied(row, destination)) {
        await hold(env, db, row, "customer_replied");
        summary.held += 1;
        continue;
      }

      // #237: before the gates, because this is not a gate — it is the message
      // no longer being wanted. Running the plan/quota checks on a reminder for
      // a job that finished yesterday would be asking permission to send
      // something nobody should send.
      //
      // FAILED rather than held: a finished job does not un-finish, so there is
      // nothing to resume. `scheduledReasonRecovers` says the same, and the two
      // must agree or the row retries forever against a condition that will
      // never change.
      if (!(await jobStillBooked(db, row))) {
        await fail(env, db, row, "job_no_longer_scheduled");
        summary.failed += 1;
        continue;
      }

      // #245/D137: a reminder contains a literal appointment time. For a task
      // mirrored to a provider, only a recent successful round trip makes that
      // time safe to repeat to the customer. This is a HOLD, not a failure:
      // the five-minute pull loop can verify the connection and the next sweep
      // will retry it without anybody recreating the reminder.
      if (!(await linkedCalendarsVerifiedRecently(db, row, now))) {
        await hold(env, db, row, "calendar_unverified");
        summary.held += 1;
        continue;
      }

      let clearance;
      try {
        clearance = await runPreSendGates(env, row.company_id, destination.to);
      } catch (cause) {
        const reason = reasonFor(cause);
        if (scheduledReasonRecovers(reason)) {
          await hold(env, db, row, reason);
          summary.held += 1;
        } else {
          await fail(env, db, row, reason);
          summary.failed += 1;
        }
        continue;
      }

      // The intent becomes a real queued message and closes, in one statement.
      // A cancel that landed while the gates were running wins here: the row is
      // no longer claimable and this returns 'gone' without writing a message.
      const fired = unwrap<FireScheduledMessageResult>(
        await db.rpc("api_fire_scheduled_message", {
          p_id: row.id,
          p_segments_estimate: Math.max(1, estimateSegments(row.body).segments),
        }),
        "fire scheduled message",
      );
      // The SQL statement repeats the two task/calendar checks while it owns
      // the scheduled row lock. A webhook, provider pull, or local task edit
      // can land after the read-side checks above; treating an atomic hold as
      // merely "not fired" would leave the owner unaware, while calling
      // `hold()` again would race the state transition SQL already made.
      if (fired.outcome === "held") {
        if (fired.reason_key !== "calendar_unverified") {
          throw new Error(
            `fire scheduled message returned unsupported hold reason ${String(fired.reason_key)}`,
          );
        }
        await disclose(
          env,
          db,
          fired.scheduled_message ?? row,
          "calendar_unverified",
        );
        summary.held += 1;
        continue;
      }
      if (fired.outcome === "failed") {
        if (fired.reason_key !== "job_no_longer_scheduled") {
          throw new Error(
            `fire scheduled message returned unsupported failure reason ${String(fired.reason_key)}`,
          );
        }
        await disclose(
          env,
          db,
          fired.scheduled_message ?? row,
          "job_no_longer_scheduled",
        );
        summary.failed += 1;
        continue;
      }
      if (fired.outcome === "gone") continue;
      if (fired.outcome !== "fired" || !fired.message) {
        throw new Error("fire scheduled message returned an invalid result");
      }

      // From here the message row exists as 'queued'. If this throws,
      // job:retry-interrupted-sends owns it — that is why the insert is
      // deliberately not rolled back on a dispatch failure.
      await dispatchOutbound(env, db, fired.message, {
        from: destination.from,
        to: destination.to,
        text: row.body,
        mediaUrls: [],
        clearance,
      });
      summary.sent += 1;
    } catch (cause) {
      failures.push(
        `${row.id}: ${cause instanceof Error ? cause.message : String(cause)}`,
      );
    }
  }

  if (failures.length > 0) {
    throw new Error(
      `scheduled send: ${failures.length} of ${due.length} failed — ${failures.join("; ")}`,
    );
  }
  return summary;
}
